### PR TITLE
Add merging of names in last-name-first style

### PR DESCRIPTION
**Pull request by mlutze**
I added a flag to `MergeNameParts` so the user can decide whether to use first-name-first or last-name-first.
I made last-name-first the default.

I slightly modified one existing test group to support the flag.

I added a new test group `test_merge_last_name_first_inverse` which ensures that merging with commas is the inverse of splitting. I skip this test for cases including a `/`, since this property doesn't hold for some of those cases.

I also had to do a bit of strange stuff with the typing imports. The flag I added is a union of literal strings, but literal string types aren't available in Python 3.7. So I try importing a literal type and if that doesn't work then I just say the type of the flag is "str", which is too general but the closest I can get

### CODE REVIEW

**Comment by mlutze** [comment]
> Thanks for your work on this. Just to avoid a misunderstanding: I wont review this PR while it is marked as draft.

Understood. There are still some changes to testing and documentation I'd like to make before submitting for review.

**Comment by mlutze** [comment]
@MiWeiss This is ready for review.

File: bibtexparser/middlewares/names.py [hunk 1]
```diff
@@ -7,6 +7,14 @@
 import dataclasses
 from typing import List, Tuple
 
+# Literal is only available in Python 3.8+
+try:
```
Comment (MiWeiss) [inline bibtexparser/middlewares/names.py:11]: In the meantime, python 3.7 has reached EOL. 

I have just updated the CI on master to not use 3.7 anymore, thus feel free to remove your fallback.

File: tests/middleware_tests/test_names.py [hunk 2]
```diff
@@ -860,6 +860,27 @@ def test_split_name_into_parts(name, expected_as_dict, strict):
...
+@pytest.mark.parametrize("strict", [True, False], ids=["strict", "non-strict"])
+def test_merge_last_name_first_inverse(name, expected_as_dict, strict):
+    """Tests that merging name parts using the last-name-first method
+    maintains the "semantics" of the name.
+
+    This property does not hold for certain values that contain '\\'.
+    """
+
+    # skip cases with \
+    if any("\\" in "".join(name) for name in expected_as_dict.values()):
```
Comment (MiWeiss) [inline tests/middleware_tests/test_names.py:875]: Can we somehow document in which cases we cannot handle this? I have struggle understanding the severity of this case.

**Review by MiWeiss — CHANGES_REQUESTED** [review]
Hi @mlutze 

Thanks a lot for opening th e PR. This is certainly a very important contribution.

I see (and acknowledge) that the whole `\\` handling is a bit of a struggle. Still, I'd like to understand it a bit better to know whats going on.  I hope you don't mind my corresponding review ;-)

File: tests/middleware_tests/test_names.py (hunk 2 above)
Comment (mlutze) [reply tests/middleware_tests/test_names.py:875]: As far as I can tell, the cases that cannot be handled are those where either the last name or "von" part ends with an odd number of `\`, since in those cases the `,` is escaped when the name parts are put back together. I add an extra `\` in those cases, which I _think_ preserves the semantics of the name, though I'm not 100% sure.

The filter here is more aggressive than what I put above though, to keep it simple. But I guess it's better if the filter captures exactly the cases for which the "inverse" property holds?

Where would you like the documentation for the edge cases?

Comment (MiWeiss) [reply tests/middleware_tests/test_names.py:875]: Ok, I guess that's enough of an edge case and ok to have it in the code, since I really do not think that single trailing `\`  are common enough to block this PR :-)

Comment (MiWeiss) [reply tests/middleware_tests/test_names.py:875]:
```suggestion
    # cases where either the last name or "von" part ends with an odd number of `\` cannot be handled, 
    # since in those cases the `,` is escaped when the name parts are put back together
    if any("\\" in "".join(name) for name in expected_as_dict.values()):
```

Comment (mlutze) [reply tests/middleware_tests/test_names.py]: Thanks. I've added the comment and made the filter a bit more lenient so only cases with an odd number of trailing `\` are discarded.

**Comment by MiWeiss** [comment]
Thanks again @mlutze  for your PR, and sorry that the review took so long!

**Review by MiWeiss — APPROVED** [review]
